src/sinks/nats.rs

```diff
@@ -0,0 +1,152 @@
+use crate::{
+    config::{DataType, GenerateConfig, SinkConfig, SinkContext, SinkDescription},
+    emit,
+    event::Event,
+    internal_events::NatsEventSent,
+    sinks::util::StreamSink,
+};
+use async_trait::async_trait;
+use futures::{future, stream::BoxStream, FutureExt, StreamExt};
+use serde::{Deserialize, Serialize};
+
+/**
+ * Code dealing with the SinkConfig struct.
+ *
+ * DEV: Start with the bare minimum for now.
+ */
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct NatsSinkConfig {
+    url: String,
+    subject: String,
+}
+
+inventory::submit! {
+    SinkDescription::new::<NatsSinkConfig>("nats")
+}
+
+impl GenerateConfig for NatsSinkConfig {}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "nats")]
+impl SinkConfig for NatsSinkConfig {
+    async fn build(
+        &self,
+        _cx: SinkContext,
+    ) -> crate::Result<(super::VectorSink, super::Healthcheck)> {
+        let sink = NatsSink::new(self.clone());
+        let healthcheck = future::ok(()).boxed();
+
+        Ok((super::VectorSink::Stream(Box::new(sink)), healthcheck))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "nats"
+    }
+}
+
+/**
+ * Code dealing with the Sink struct.
+ */
+
+#[derive(Clone)]
+pub struct NatsSink {
+    nc: nats::Connection,
+    subject: String,
+}
+
+impl NatsSink {
+    fn new(config: NatsSinkConfig) -> Self {
+        let options = nats::Options::new();
+        let nc = options.connect(&config.url).unwrap();
+
+        Self {
+            nc,
+            subject: config.subject,
+        }
+    }
+}
+
+#[async_trait]
+impl StreamSink for NatsSink {
+    async fn run(&mut self, mut input: BoxStream<'_, Event>) -> Result<(), ()> {
+        while let Some(event) = input.next().await {
+            match event {
+                Event::Log(log) => {
+                    let body = log
+                        .get(crate::config::log_schema().message_key())
+                        .map(|v| v.to_string_lossy())
+                        .unwrap_or_else(|| "".into());
+
+                    let message_len = body.len();
+
+                    if self.nc.publish(&self.subject, body).is_ok() {
+                        emit!(NatsEventSent {
+                            byte_size: message_len,
+                        });
+                    }
+                }
+                Event::Metric(_metric) => {}
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(feature = "nats-integration-tests")]
+#[cfg(test)]
+mod integration_tests {
+    use super::*;
+    use crate::test_util::{random_lines_with_stream, random_string, trace_init};
+    use std::{
+        sync::{Arc, Mutex},
+        time::Duration,
+    };
+
+    #[tokio::test]
+    async fn nats_happy() {
+        // Publish `N` messages to NATS.
+        //
+        // Observe with a second subscriber that at least some of the messages
```

Comment:
This is the only subscriber, no? `vector` would be the publisher here.

I realize `nats` is "at most once", but if we start the subscriber first, we should receive all messages, no? It's just if the subscriber is offline that messages would be dropped.

Reply:
Yeah, it's the only subscriber with Vector being the publisher like you said. I think I had meant to say "separate."

We should, I was just overly conservatively estimating conditions that might make the integration test flaky.

Reply:
This test seems to be passing for me now (and I note that it verifies all messages were sent and received).